### DISCUSSION TITLE
assign membership to SSO customers

### DISCUSSION THREAD

**Discussion by DHD-Harper** [pretix/pretix#3455] [Ideas]
It would be helpful if customers who have been created through a SSO integration can be given a membership automatically so they can purchase reserved products.

**Comment by raphaelm** [comment]
Yes, that would be great. But we haven't yet come up with a flexible concept around this. In your case, would *every* SSO user get the membership? Where would we derive the type of membership and expiration etc from?

**Reply by jasle** [reply]
My thoughts and ideas on this:
For determining which type of membership should be given I see two cases: There is one type of membership, that should be given to everyone who can sign in using an SSO method or getting the type of the membership from a field in OpenID Connect.
So it would need something like an Internal identifier for membership types and the options to configure either a type of membership or the field to get the type of membership from when setting up an SSO provider.

For the expiration I also see two cases: The membership is valid as long as a login using SSO works or having a field in OpenID Connect with the expiration date.
So it would need the options to configure either 'always' or a field to get the expiration date from when setting up an SSO provider.

**Reply by jasle** [reply]
Would you be open for a PR implementing this with the options above?

**Reply by raphaelm** [reply]
The problem for me is that OpenID Connect defines no standardized way to signify a group membership that I'm aware of and there are also no conventions in the industry that I know of. Most people do not control their IdP in the way that they can programatically add information to the profile response, so for many people it will alway need a second API request.

Since our OIDC implementation strives to be very standard-compliant and universally usable, I think we'd only merge a PR if it starts with some research into what options various big IdPs (think Keycloak, Okta, …) provide to be used for such a thing and then we could discuss what configuration would be needed inside pretix to utilize that.

Otherwise, extending the plugin API is likely the better choice.

**Comment by DHD-Harper** [comment]
Yes, everyone with SSO should get a specific membership. There may be multiple SSO providers, each of which should lead to an associated membership. Membership shall expire only when there is no longer an active SSO for the user.

An active SSO at a specific provider is sole proof for a specific membership.

**Reply by raphaelm** [reply]
I see. These answers make sense, but they might be different for everyone else who needs this, and I'm a little bit horrified of the complexity required to make this configurable properly ;)

**Comment by DHD-Harper** [comment]
It would be also helpful if customers are only allowed to place _one_ order per event as part of their SSO associated membership.

**Reply by raphaelm** [reply]
That would be supported through MembershipType.allow_parallel_usage

**Comment by kiancross** [comment]
Are any signals currently dispatched when a user signs in/account creation? I couldn't see one in the documentation, but it would allow something like this to be implemented as a plugin, which might be slightly better given the complex configurability issues?

**Reply by raphaelm** [reply]
Not yet, but I think it would be a good idea!

(With the caveat that we don't have a concept of plugins being enabled and disabled on organizer level currently, so the plugin would need to take care of that itself through settings)
